File: src/main/java/de/lukaseichler/recurrentneuralnetwork/RecurrentNeuralNetwork.java
package de.lukaseichler.recurrentneuralnetwork;

import java.util.List;
import javax.annotation.Nonnull;

/**
 * @author leichler
 */
public class RecurrentNeuralNetwork extends Network {

    private List<Double> previousState;

    @Override public List<Double> calculate(@Nonnull final List<Double> input) {
        List<Double> result = super.calculate(input);
        for (int i = 0; i < result.size(); i++) {
            result.set(i, result.get(i) * getState(i));
        }
        previousState = result;
        return result;
    }

    private Double getState(final int i) {
        if (previousState == null) {
            return 1.0;
        }
        return previousState.get(i);
    }
}

File: src/main/scala/de/lukaseichler/recurrentneuralnetwork/RecurrentNeuralNetwork.scala
package de.lukaseichler.recurrentneuralnetwork

import javax.annotation.Nonnull

/**
 * @author Lukas Eichler
 */
class RecurrentNeuralNetwork extends Network {
  private var previousState: List[Double] = null

  override def calculate(@Nonnull input: List[Double]): List[Double] = {
    val result: List[Double] = super.calculate(input)
    input.zipWithIndex.foreach { case (x, i) => result.updated(i, result(i) * getState(i))}
    previousState = result
    result
  }

  private def getState(i: Int): Double = {
    if (previousState == null) {
      return 1.0
    }

    previousState(i)
  }
}
